### PR TITLE
[FLINK-34948][cdc-common] Fix cdc `RowType` can not convert to flink type

### PR DESCRIPTION
**Pull request by zhongqishang**
Fix cdc `RowType` can not convert to flink type

I meet the follow exception:
```
java.lang.ArrayStoreException
	at java.lang.System.arraycopy(Native Method)
	at java.util.Arrays.copyOf(Arrays.java:3213)
	at java.util.ArrayList.toArray(ArrayList.java:413)
	at java.util.Collections$UnmodifiableCollection.toArray(Collections.java:1036)
```

### CODE REVIEW

**Comment by PatrickRen** [comment]
@zhongqishang Thanks for the PR! Could you open a Jira issue for this one? Thanks

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import static org.apache.flink.table.api.DataTypes.TIME;
+import static org.apache.flink.table.api.DataTypes.TIMESTAMP_LTZ;
+import static org.apache.flink.table.api.DataTypes.TIMESTAMP_WITH_TIME_ZONE;
+import static org.apache.flink.table.api.DataTypes.TINYINT;
+import static org.apache.flink.table.api.DataTypes.VARBINARY;
+import static org.apache.flink.table.api.DataTypes.VARCHAR;
+import static org.junit.Assert.assertEquals;
+
+/** A test for the {@link org.apache.flink.cdc.common.types.utils.DataTypeUtils}. */
+public class DataTypeUtilsTest {
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:56]: Thanks @zhongqishang for this hotfix. 
Looks overall good to me.
Could you help change the test from JUnit4 to JUnit5 and minimize the visibility for class and methods?

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+    private final DataType[] allTypes =
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:57]: nit: can be const.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import static org.apache.flink.table.api.DataTypes.ROW;
+import static org.apache.flink.table.api.DataTypes.SMALLINT;
+import static org.apache.flink.table.api.DataTypes.STRING;
+import static org.apache.flink.table.api.DataTypes.TIME;
+import static org.apache.flink.table.api.DataTypes.TIMESTAMP_LTZ;
+import static org.apache.flink.table.api.DataTypes.TIMESTAMP_WITH_TIME_ZONE;
+import static org.apache.flink.table.api.DataTypes.TINYINT;
+import static org.apache.flink.table.api.DataTypes.VARBINARY;
+import static org.apache.flink.table.api.DataTypes.VARCHAR;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:53]:
```suggestion
import static org.assertj.core.api.Assertions.assertThat;
```

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+/** A test for the {@link org.apache.flink.cdc.common.types.utils.DataTypeUtils}. */
+public class DataTypeUtilsTest {
+    private static final DataType[] allTypes =
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:57]:
```suggestion
    private static final DataType[] ALL_TYPES =
```

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java [hunk 5]
```diff
@@ -26,10 +26,12 @@
...
 import org.apache.flink.cdc.common.types.DataTypes;
+import org.apache.flink.cdc.common.types.RowType;
 import org.apache.flink.cdc.common.utils.Preconditions;
 import org.apache.flink.util.CollectionUtil;
 
 import java.util.List;
+import java.util.stream.Collectors;
 
 /** Utilities for handling {@link DataType}s. */
 public class DataTypeUtils {
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:37]:
```suggestion 
class DataTypeUtils {
```

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                DataTypes.ARRAY(DataTypes.BIGINT()),
+                DataTypes.MAP(DataTypes.SMALLINT(), DataTypes.STRING()),
+                DataTypes.ROW(
+                        DataTypes.FIELD("f1", DataTypes.STRING()),
+                        DataTypes.FIELD("f2", DataTypes.STRING(), "desc")),
+                DataTypes.ROW(DataTypes.SMALLINT(), DataTypes.STRING())
+            };
+
+    @Test
+    public void testToFlinkDataType() {
```
Comment (Jiabao-Sun) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:91]:
```suggestion
    void testToFlinkDataType() {
```

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 5 above)
Comment (zhongqishang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java]: I get the error `Test class 'DataTypeUtilsTest' is not constructable because it is not 'public'`.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java (hunk 6 above)
Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:91]: I get the error `Method 'testToFlinkDataType' annotated with '@Test' should be public `

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java (hunk 1 above)
Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:56]: Done.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java (hunk 2 above)
Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java]: Done.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 5 above)
Comment (zhongqishang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:37]: Changing the scope will be compile failed.

**Comment by zhongqishang** [comment]
@Jiabao-Sun Thanks for your detailed review, all fixed.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java [hunk 7]
```diff
@@ -134,8 +136,27 @@ public static org.apache.flink.table.types.DataType toFlinkDataType(DataType typ
...
+                                                                                dataField
+                                                                                        .getType()))
+                                                        : org.apache.flink.table.api.DataTypes
+                                                                .FIELD(
+                                                                        dataField.getName(),
+                                                                        toFlinkDataType(
+                                                                                dataField
+                                                                                        .getType()),
+                                                                        dataField.getDescription()))
+                                .collect(Collectors.toList());
```
Comment (laglangyue) [inline flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:158]: the lambda is so ugly, maybe you can define a map function

Comment (zhongqishang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:144]: Add a named `toFlinkDataTypeField` method.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 5 above)
Comment (Jiabao-Sun) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:38]: Sorry, I marked it on the wrong class.
It should be `DataTypeUtilsTest`.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java (hunk 6 above)
Comment (Jiabao-Sun) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java:91]: In JUnit5, test  class and methods do not need to be declared as public.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 7 above)
Comment (laglangyue) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:144]: In my opinion, `toFlinkDataTypeField` is a method of DataField, not a static function in utils.
Hope for a idea from Project member.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 5 above)
Comment (zhongqishang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:38]: Get it, Done.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java (hunk 6 above)
Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/types/utils/DataTypeUtilsTest.java]: Get it, Done.

File: flink-cdc-common/src/test/java/org/apache/flink/cdc/common/utils/SchemaUtilsTest.java [hunk 8]
```diff
@@ -27,20 +27,21 @@
 import org.apache.flink.cdc.common.types.DataType;
 import org.apache.flink.cdc.common.types.DataTypes;
 
-import org.junit.Assert;
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
```
Comment (laglangyue) [inline flink-cdc-common/src/test/java/org/apache/flink/cdc/common/utils/SchemaUtilsTest.java:30]: It change SchemaUtilsTest IT to Junit5, it seems not related to this PR. And we separate a PR to do it in usually.

Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/utils/SchemaUtilsTest.java:30]: OK, I will be separate a new PR.

Comment (zhongqishang) [reply flink-cdc-common/src/test/java/org/apache/flink/cdc/common/utils/SchemaUtilsTest.java]: Rollback.

File: flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java (hunk 7 above)
Comment (zhongqishang) [reply flink-cdc-common/src/main/java/org/apache/flink/cdc/common/types/utils/DataTypeUtils.java:144]: Yes, this will be better, Fixed.

**Review by Jiabao-Sun — APPROVED** [review]
Thanks @zhongqishang for this hotfix and @laglangyue for the review.
LGTM.
